typescript/schedules.md

```diff
@@ -0,0 +1,1246 @@
+   */
+  TerminateOther,
+
+  /**
+   * Allow any number of Actions to start immediately.
+   *
+   * This is the only policy under which multiple Actions can run concurrently.
+   *
+   * {@link lastCompletionResult} and {@link lastFailure} will not be available.
+   */
+  AllowAll,
+}
+
+export interface Backfill {
+  start: Date;
+  end: Date;
+  /**
+   * Override overlap policy for this request.
+   */
+  overlap?: ScheduleOverlapPolicy;
+}
+
+/**
+ * The range of values depends on which, if any, fields are set:
+ *
+ * ```
+ * {} -> all values
+ * {start} -> start
+ * {start, end} -> every value from start to end (implies step = 1)
+ * {start, step} -> from start, by step (implies end is max for this field)
+ * {start, end, step} -> from start to end, by step
+ * {step} -> all values, by step (implies start and end are full range for this field)
+ * {end} and {end, step} are not allowed
+ * ```
+ * TODO is there a way to express not allowed ^ in type?
+ *
+ * For example:
+ *
+ * ```
+ * {start: 2, end: 10, step: 3} -> 2, 5, 8
+ * ```
+ */
+export interface Range<Unit> {
+  start?: Unit;
+  end?: Unit;
+
+  /**
+   * The step to take between each value.
+   *
+   * @default 1
+   */
+  step?: PositiveInteger;
+}
+
+export type AnyValue = '*';
+
+export type ZeroTo59 =
+  | 0
+  | 1
+  | 2
+  | 3
+  | 4
+  | 5
+  | 6
+  | 7
+  | 8
+  | 9
+  | 10
+  | 11
+  | 12
+  | 13
+  | 14
+  | 15
+  | 16
+  | 17
+  | 18
+  | 19
+  | 20
+  | 21
+  | 22
+  | 23
+  | 24
+  | 25
+  | 26
+  | 27
+  | 28
+  | 29
+  | 30
+  | 31
+  | 32
+  | 33
+  | 34
+  | 35
+  | 36
+  | 37
+  | 38
+  | 39
+  | 40
+  | 41
+  | 42
+  | 43
+  | 44
+  | 45
+  | 46
+  | 47
+  | 48
+  | 49
+  | 50
+  | 51
+  | 52
+  | 53
+  | 54
+  | 55
+  | 56
+  | 57
+  | 58
+  | 59;
+
+export type Second = Range<ZeroTo59> | ZeroTo59;
+export type Minute = Range<ZeroTo59> | ZeroTo59;
+
+export type ZeroTo23 =
+  | 0
+  | 1
+  | 2
+  | 3
+  | 4
+  | 5
+  | 6
+  | 7
+  | 8
+  | 9
+  | 10
+  | 11
+  | 12
+  | 13
+  | 14
+  | 15
+  | 16
+  | 17
+  | 18
+  | 19
+  | 20
+  | 21
+  | 22
+  | 23;
+
+export type Hour = Range<ZeroTo23> | ZeroTo23;
+
+export type ZeroTo30 =
+  | 0
+  | 1
+  | 2
+  | 3
+  | 4
+  | 5
+  | 6
+  | 7
+  | 8
+  | 9
+  | 10
+  | 11
+  | 12
+  | 13
+  | 14
+  | 15
+  | 16
+  | 17
+  | 18
+  | 19
+  | 20
+  | 21
+  | 22
+  | 23
+  | 24
+  | 25
+  | 26
+  | 27
+  | 28
+  | 29
+  | 30;
+
+export type DayOfMonth = Range<ZeroTo30> | ZeroTo30;
+
+export type ZeroTo11 = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11;
+
+export type MonthName =
+  | 'JAN'
+  | 'JANUARY'
+  | 'FEB'
+  | 'FEBRUARY'
+  | 'MAR'
+  | 'MARCH'
+  | 'APR'
+  | 'APRIL'
+  | 'MAY'
+  | 'JUN'
+  | 'JUNE'
+  | 'JUL'
+  | 'JULY'
+  | 'AUG'
+  | 'AUGUST'
+  | 'SEP'
+  | 'SEPTEMBER'
+  | 'OCT'
+  | 'OCTOBER'
+  | 'NOV'
+  | 'NOVEMBER'
+  | 'DEC'
+  | 'DECEMBER';
+
+export type MonthValue = ZeroTo11 | MonthName;
+
+export type Month = Range<MonthValue> | MonthValue;
+
+export type TwoThousandTo2100 =
```

Comment:
oh wow, I know you people like your types, but imo this doesn't add any value :slightly_smiling_face:   the year bounds are an arbitrary limitation for performance reasons, and easy to change, not something that should be encoded in a type system

Reply:
If user specifies year 2101, will they get an error, or will it work later on when server increases time range?

Reply:
What if the user specifies `numberFromUserInput` variable? What if I'm using JavaScript?

We need to stop pretending these types are for literals and hand-typed code. Advanced validation is a runtime concern. TypeScript doesn't have dependent types to properly enforce this at compile time, so this isn't giving much benefit, just blowing up the definition size.

Reply:
Yes, this is overdoing it IMO.

Reply:
no error, it just won't match those times unless server is updated

Reply:
For years, planning on using `number` and not validating. Let them use 2101, and server will be updated by then.